langchain/src/callbacks/index.ts

```diff
@@ -0,0 +1,259 @@
+import { AgentAction, AgentFinish } from "../agents/index.js";
+import { ChainValues } from "../chains/index.js";
+import { LLMResult } from "../schema/index.js";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Error = any;
+
+export interface BaseCallbackHandlerInput {
+  alwaysVerbose?: boolean;
+  ignoreLLM?: boolean;
+  ignoreChain?: boolean;
+  ignoreAgent?: boolean;
+}
+
+export abstract class BaseCallbackHandler implements BaseCallbackHandlerInput {
+  alwaysVerbose = false;
+
+  ignoreLLM = false;
+
+  ignoreChain = false;
+
+  ignoreAgent = false;
+
+  constructor(input?: BaseCallbackHandlerInput) {
+    if (input) {
+      this.alwaysVerbose = input.alwaysVerbose ?? this.alwaysVerbose;
+      this.ignoreLLM = input.ignoreLLM ?? this.ignoreLLM;
+      this.ignoreChain = input.ignoreChain ?? this.ignoreChain;
+      this.ignoreAgent = input.ignoreAgent ?? this.ignoreAgent;
+    }
+  }
+
+  handleLLMStart?(
+    llm: { name: string },
+    prompts: string[],
+    verbose?: boolean
+  ): Promise<void>;
+
+  handleLLMNewToken?(token: string, verbose?: boolean): Promise<void>;
+
+  handleLLMError?(err: Error, verbose?: boolean): Promise<void>;
+
+  handleLLMEnd?(output: LLMResult, verbose?: boolean): Promise<void>;
+
+  handleChainStart?(
+    chain: { name: string },
+    inputs: ChainValues,
+    verbose?: boolean
+  ): Promise<void>;
+
+  handleChainError?(err: Error, verbose?: boolean): Promise<void>;
+
+  handleChainEnd?(outputs: ChainValues, verbose?: boolean): Promise<void>;
+
+  handleToolStart?(
+    tool: { name: string },
+    input: string,
+    verbose?: boolean
+  ): Promise<void>;
+
+  handleToolError?(err: Error, verbose?: boolean): Promise<void>;
+
+  handleToolEnd?(output: string, verbose?: boolean): Promise<void>;
+
+  handleText?(text: string, verbose?: boolean): Promise<void>;
```

Comment:
Remind me when this one is fired?

Reply:
This is used whenever we need to log arbitrary messages from chains or agents that are not necessarily on start/end events. TBH I'm not in love with this, and want to look into another soln but it's not a priority rn, so I just copied what's in python